src/type-layout.md

```diff
@@ -86,9 +86,10 @@ String slices are a UTF-8 representation of characters that have the same layout
 
 ## Tuple Layout
 
-Tuples do not have any guarantees about their layout.
+Tuples have the same layout guarantees as a struct with the same fields when
```

Comment:
I'm opposed to documenting this, as it precludes us from tuple variadic approaches which require "tail borrowing."

Reply:
@rfcbot concern tuple-tail-borrowing

Reply:
Can you elaborate on what that is?

Reply:
There have been a number of discussions in relation to variadic generics about wanting the ability to write code which could recurse over the structure of a tuple, HList-style. In order for this to work, the "tail" of a tuple must be represented contiguously. That is, for tuple `(A, B, C, D)`, there must be a sub-tuple `(B, C, D)` in memory. Allowing arbitrary ordering as structs do would cause layouts like `(A, D, C, B)`, from which it's impossible to create a reference to a tuple of a sub-structure like `(B, C, D)`.

Note that hlist-compatible representations of tuples need not be strictly ordered-- it's fine for each new element to be placed at either the front, back, or even inside of the tuple so long as the tail tuple remains a part of the representation.

I'd like our representation documentation to stay forward-compatible with tuple tail-borrowing until we've explicitly decided not to do it so that users aren't expecting to be able to transmute between `struct S(A, B, C, D);` and tuple `(A, B, C, D)`.

Reply:
As I intended the guarantees I have written, what you described here would still be backwards-compatible to introduce, because that particular ordering is just one among the many different orderings that the guarantees allow the compiler to choose. As for the struct, you also can't transmute between different `#[repr(Rust)]` structs with the same fields either.

I would be happy to elaborate on this in the text though.

Reply:
I see, I misunderstood "have the same layout *guarantees* as a struct with the same fields". I over-read this to mean "have the same *layout* as a struct with the same fields". Thanks for clarifying! Still, I can imagine this documentation going stale: if we provide more guarantees for structs in the future, we wouldn't necessarily match those guarantees for tuples.

Reply:
I tend to think that tuples and tuple structs ought to have analogous layout. The UCG conversation at least came to that conclusion (we do discuss "tail references" as well). There was talk, as I recall, of making `(T, T, T)` sort of special case, to permit casting that to `[T; 3]`.

Reply:
What do you think of the new phrasing?